### ISSUE TITLE
No longer works because google changed their implementation

### ISSUE THREAD

**Issue by danieleisenhardt** [dpa99c/cordova-plugin-cloud-settings#17]
From the page the link in the docs redirects to:
> Note: For most apps, Auto Backup should be used to implement backup and restore. Your app can only implement Auto Backup or key-value backup, not both. Auto Backup requries no code and will back up entire files while key-value backup requires you write code to define your backup content explicitly in the form of key-value pairs.
>
> **In the past, developers registered their app to get a service key. You no longer need to register your app for key-value backup.**

Yet the install command for this plugin requires ANDROID_BACKUP_SERVICE_KEY or this error comes up:
```
$ cordova plugin add cordova-plugin-cloud-settings
Installing "cordova-plugin-cloud-settings" for android
Failed to install 'cordova-plugin-cloud-settings': Error: Variable(s) missing: ANDROID_BACKUP_SERVICE_KEY
    at Object.mergeVariables (/usr/local/lib/node_modules/cordova/node_modules/cordova-lib/src/plugman/variable-merge.js:58:15)
    at /usr/local/lib/node_modules/cordova/node_modules/cordova-lib/src/plugman/install.js:307:44
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
Variable(s) missing: ANDROID_BACKUP_SERVICE_KEY
```

**Comment by GabrielCalin** [comment]
Personally, I tried with a random API key, hoping it will work (like "123"). I managed to install the plugin and it kind of works, except that Google doesn't automatically backup my app settings. I'm using a Xiaomi phone and I am forced to manually backup my app data to a file and then manually restore it after a reinstall. Then. the cloud settings plugin comes into action and it sees the backup data and loads it. I can only assume that these manual interventions are needed because I haven't used a proper API key, but maybe I'm wrong, although I'm not sure what I'm doing wrong (I'm just following the described lifecycle, but the cloud settings are not automatically backed up).

**Comment by Brianfit** [comment]
I don't think it's your key. I have a valid key and while the plugin installs and functions according to the plugin's life cycle description on iOS, in Android the onRestore() function does not seem to detect cloud storage updates, and while I can write and read to the cloud on Android it seems the file remains device-specific. I'm curious GabrielCain whether your manual interventions are cross-device or single device only?

**Comment by GabrielCalin** [comment]
Brianfit, I think the manual backup of the app's storage is Xiaomi specific. From my understanding, the procedure creates a backup file that I could upload to a PC and restore it on another (Xiaomi) device. However, I only tested it on a single device: I backed up my app's storage, uninstalled the app, reinstalled it, restored the backup and then started the app -> at this point, the cloud settings plugin was able to load the saved settings and restore them. This is not a proper solution though.

**Comment by Brianfit** [comment]
From my logcat -- I believe from an onRestore event though it could be an onLoad as well -- I don't see any way to parse the precise line in my code which generated this. The file successfully loads what appears to be a local file, but ignores the existence of an alternative cloud version saved from another device. 
```
Attempted to send a second callback for ID: CloudSettingsPlugin955246206
    Result was: "Invalid action"
```
@dpa99c Is this plugin still being supported at all? I've never seen my onRestore callback triggered on Android, though it works ok on iOS. It's a uniquely useful plugin and it would be great to see it back in action on Android, if indeed this is a plugin issue and not a user installation error on my (our) part.

**Comment by dpa99c** [comment]
I have removed the `ANDROID_BACKUP_SERVICE_KEY` variable in v2.0.0 of this plugin which I've just published.

The error message `Attempted to send a second callback for ID: CloudSettingsPlugin955246206` is because the plugin was not correctly returning the result which has also been fixed in v2.0.0


As outlined in [this](https://stackoverflow.com/q/48873335/777265) SO post, `onRestore` on Android is only observed to be invoked if you manually call [requestRestore](https://developer.android.com/reference/android/app/backup/BackupManager#requestRestore(android.app.backup.RestoreObserver)). 
However, `requestRestore` is deprecated in Android 9.0 / API 28 and performs a no-op so in practice `onRestore` will not be invoked on Android.
However that doesn't mean that Android backup and therefore this plugin will not work - my empirical regression testing of the v2.0.0 of this plugin confirms it works correctly and the cloud settings file is restored from Android backup upon app installation and made available to the app at first run.
Note that if an automatically-invoked restore operation takes place on Android while your app is running, the app process will be terminated once the restore operation is complete.
Upon relaunching the app, any newly restored data will be made available to the app.

Therefore I believe the various points raised in this issue have been covered by the release of v2.0.0 and I'm therefore closing it.
